### ISSUE TITLE
Compiling nginx (1.11.3) with ngx_brotli failed

### ISSUE THREAD

**Issue by kn007** [google/ngx_brotli#33]
A same question like issue #32 , I do not know @Sunka how to fix it. So I submit a new issue here.

Environment: Centos 6.8 x64, gcc 4.4.7 20120313 (Red Hat 4.4.7-17), libstdc++-devel-4.4.7-17.

Nginx compiling stdout:
```
adding module in /data/src/ngx_brotli
checking for Brotli library ... not found
checking for Brotli library in /usr/local/ ... not found
checking for Brotli library in /usr/pkg/ ... not found
checking for Brotli library in /opt/local/ ... not found
./configure: error: ngx_brotli filter module requires Brotli library.
```

`$cat objs/autoconf.err` : 
```
checking for Brotli library

In file included from objs/autotest.c:4:
/usr/local/include/brotli/enc/encode.h:12:18: error: string: No such file or directory
/usr/local/include/brotli/enc/encode.h:13:18: error: vector: No such file or directory
In file included from /usr/local/include/brotli/enc/./command.h:12,
                 from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././fast_log.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./command.h:13,
                 from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././prefix.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./command.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./hash.h:14:21: error: algorithm: No such file or directory
/usr/local/include/brotli/enc/./hash.h:15:19: error: cstring: No such file or directory
/usr/local/include/brotli/enc/./hash.h:16:18: error: limits: No such file or directory
In file included from /usr/local/include/brotli/enc/./hash.h:18,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././dictionary_hash.h:14: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:20,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././find_match_length.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:23,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././static_dict.h:14: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:24,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././transform.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./hash.h:27: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:16,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./ringbuffer.h:12:40: error: cstdlib: No such file or directory
In file included from /usr/local/include/brotli/enc/encode.h:16,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./ringbuffer.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:18,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./streams.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from objs/autotest.c:4:
/usr/local/include/brotli/enc/encode.h:21: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
----------

#include <sys/types.h>
#include <unistd.h>
#include <brotli/enc/encode.h>

int main(void) {
    BrotliEncoderCreateInstance(NULL, NULL, NULL);
    return 0;
}

----------
cc -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -o objs/autotest objs/autotest.c -ljemalloc -lrt -lbrotlienc
----------

----------------------------------------
checking for Brotli library in /usr/local/

In file included from objs/autotest.c:4:
/usr/local/include/brotli/enc/encode.h:12:18: error: string: No such file or directory
/usr/local/include/brotli/enc/encode.h:13:18: error: vector: No such file or directory
In file included from /usr/local/include/brotli/enc/./command.h:12,
                 from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././fast_log.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./command.h:13,
                 from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././prefix.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./command.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./hash.h:14:21: error: algorithm: No such file or directory
/usr/local/include/brotli/enc/./hash.h:15:19: error: cstring: No such file or directory
/usr/local/include/brotli/enc/./hash.h:16:18: error: limits: No such file or directory
In file included from /usr/local/include/brotli/enc/./hash.h:18,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././dictionary_hash.h:14: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:20,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././find_match_length.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:23,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././static_dict.h:14: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:24,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././transform.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./hash.h:27: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:16,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./ringbuffer.h:12:40: error: cstdlib: No such file or directory
In file included from /usr/local/include/brotli/enc/encode.h:16,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./ringbuffer.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:18,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./streams.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from objs/autotest.c:4:
/usr/local/include/brotli/enc/encode.h:21: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
----------

#include <sys/types.h>
#include <unistd.h>
#include <brotli/enc/encode.h>

int main(void) {
    BrotliEncoderCreateInstance(NULL, NULL, NULL);
    return 0;
}

----------
cc -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -I /usr/local/include -o objs/autotest objs/autotest.c -ljemalloc -lrt
 -L/usr/local/lib -lbrotlienc
----------

----------------------------------------
checking for Brotli library in /usr/pkg/

In file included from objs/autotest.c:4:
/usr/local/include/brotli/enc/encode.h:12:18: error: string: No such file or directory
/usr/local/include/brotli/enc/encode.h:13:18: error: vector: No such file or directory
In file included from /usr/local/include/brotli/enc/./command.h:12,
                 from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././fast_log.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./command.h:13,
                 from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././prefix.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./command.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./hash.h:14:21: error: algorithm: No such file or directory
/usr/local/include/brotli/enc/./hash.h:15:19: error: cstring: No such file or directory
/usr/local/include/brotli/enc/./hash.h:16:18: error: limits: No such file or directory
In file included from /usr/local/include/brotli/enc/./hash.h:18,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././dictionary_hash.h:14: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:20,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././find_match_length.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:23,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././static_dict.h:14: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:24,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././transform.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./hash.h:27: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:16,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./ringbuffer.h:12:40: error: cstdlib: No such file or directory
In file included from /usr/local/include/brotli/enc/encode.h:16,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./ringbuffer.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:18,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./streams.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from objs/autotest.c:4:
/usr/local/include/brotli/enc/encode.h:21: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
----------

#include <sys/types.h>
#include <unistd.h>
#include <brotli/enc/encode.h>

int main(void) {
    BrotliEncoderCreateInstance(NULL, NULL, NULL);
    return 0;
}

----------
cc -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -I /usr/pkg/include -o objs/autotest objs/autotest.c -ljemalloc -lrt -
L/usr/pkg/lib -lbrotlienc
----------

----------------------------------------
checking for Brotli library in /opt/local/

In file included from objs/autotest.c:4:
/usr/local/include/brotli/enc/encode.h:12:18: error: string: No such file or directory
/usr/local/include/brotli/enc/encode.h:13:18: error: vector: No such file or directory
In file included from /usr/local/include/brotli/enc/./command.h:12,
                 from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././fast_log.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./command.h:13,
                 from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././prefix.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:14,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./command.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./hash.h:14:21: error: algorithm: No such file or directory
/usr/local/include/brotli/enc/./hash.h:15:19: error: cstring: No such file or directory
/usr/local/include/brotli/enc/./hash.h:16:18: error: limits: No such file or directory
In file included from /usr/local/include/brotli/enc/./hash.h:18,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././dictionary_hash.h:14: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:20,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././find_match_length.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:23,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././static_dict.h:14: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/./hash.h:24,
                 from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/././transform.h:16: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:15,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./hash.h:27: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:16,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./ringbuffer.h:12:40: error: cstdlib: No such file or directory
In file included from /usr/local/include/brotli/enc/encode.h:16,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./ringbuffer.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from /usr/local/include/brotli/enc/encode.h:18,
                 from objs/autotest.c:4:
/usr/local/include/brotli/enc/./streams.h:17: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
In file included from objs/autotest.c:4:
/usr/local/include/brotli/enc/encode.h:21: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'brotli'
----------

#include <sys/types.h>
#include <unistd.h>
#include <brotli/enc/encode.h>

int main(void) {
    BrotliEncoderCreateInstance(NULL, NULL, NULL);
    return 0;
}

----------
cc -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -I /opt/local/include -o objs/autotest objs/autotest.c -ljemalloc -lrt
 -L/opt/local/lib -lbrotlienc
----------
```

I saw that have two files could not include in `/usr/local/include/brotli/enc/encode.h`.
```
error: string: No such file or directory 
error: vector: No such file or directory
```

`$cat /usr/local/include/brotli/enc/encode.h`(Line 9~19) : 
```
#ifndef BROTLI_ENC_ENCODE_H_
#define BROTLI_ENC_ENCODE_H_

#include <string>
#include <vector>
#include "./command.h"
#include "./hash.h"
#include "./ringbuffer.h"
#include "./static_dict.h"
#include "./streams.h"
#include "./types.h"
```

How can i fix that?

Thank you.

**Comment by redburn** [comment]
Same issue here, same OS. Builds just fine on CentOS 7.

**Comment by PiotrSikora** [comment]
This is incompatibility between very old [libbrotli](https://github.com/bagder/libbrotli) and latest [ngx_brotli](https://github.com/google/ngx_brotli).

You need to either use recent [libbrotli](https://github.com/bagder/libbrotli) with C interface (https://github.com/bagder/libbrotli/commit/3f46e28f70ba86f222b7f7d87a8b677ab84239ae from April, or newer) or use older [ngx_brotli](https://github.com/google/ngx_brotli) written in C++ (https://github.com/google/ngx_brotli/commit/ada972b2d4492d911d0ab7c9d467df8e819d8247 or older). There was transition period where [Brotli](https://github.com/google/brotli) exposed both: C and C++ interfaces and where [ngx_brotli](https://github.com/google/ngx_brotli) would work with either one, but that period has ended and both are C-only now.

**Comment by kn007** [comment]
Thank you.
I am using git-clone to get the latest libbrotli and latest ngx_brotli, but still have this problem.
I will test again later.

**Comment by PiotrSikora** [comment]
Did you actually install [libbrotli](https://github.com/bagder/libbrotli), i.e. did you call `make install`? The errors you pasted suggest that the headers under `/usr/local/include/brotli` are from the old C++ version of the [Brotli](https://github.com/google/brotli) library.

**Comment by redburn** [comment]
I also cloned the git repository, and followed the build instructions, including `make install`. However, I just repeated all the steps and now it _is_ working. The headers in `/usr/local/include/brotli` are different than before, though I'm pretty sure I ran the same commands and used the same repository -- and @kn007 apparently had the same.

Anyway, it's working: `content-encoding: br`. Hope it works for you, too, @kn007 .
